x/staking/keeper/liquid_stake_test.go

```diff
@@ -0,0 +1,514 @@
+	total := sdk.NewInt(100)
+	app.StakingKeeper.SetTotalLiquidStakedTokens(ctx, total)
+
+	// Confirm it was updated
+	require.Equal(t, total, app.StakingKeeper.GetTotalLiquidStakedTokens(ctx), "initial")
+}
+
+// Tests Increase/Decrease TotalValidatorTotalLiquidShares
+func TestValidatorTotalLiquidShares(t *testing.T) {
+	_, app, ctx := createTestInput(t)
+
+	// Create a validator address
+	privKey := secp256k1.GenPrivKey()
+	pubKey := privKey.PubKey()
+	valAddress := sdk.ValAddress(pubKey.Address())
+
+	// Set an initial total
+	initial := sdk.NewDec(100)
+	validator := types.Validator{
+		OperatorAddress:   valAddress.String(),
+		TotalLiquidShares: initial,
+	}
+	app.StakingKeeper.SetValidator(ctx, validator)
+}
+
+// Tests AccountIsLiquidStakingProvider
+func TestAccountIsLiquidStakingProvider(t *testing.T) {
+	_, app, ctx := createTestInput(t)
+
+	// Create base account
+	baseAccountAddress := sdk.AccAddress("base-account")
+	app.AccountKeeper.SetAccount(ctx, authtypes.NewBaseAccountWithAddress(baseAccountAddress))
+
+	// Create an ICA module account
+	icaModuleAccountName := "ica-account"
+	icaAccountAddress := address.Module(icaModuleAccountName, []byte("ica-module-account"))
+	icaAccount := authtypes.NewModuleAccount(
+		authtypes.NewBaseAccountWithAddress(icaAccountAddress),
+		icaModuleAccountName,
+	)
+	app.AccountKeeper.SetAccount(ctx, icaAccount)
+
+	// Only the ICA module account should be considered a liquid staking provider
+	require.False(t, app.StakingKeeper.AccountIsLiquidStakingProvider(ctx, baseAccountAddress), "base account")
+	require.True(t, app.StakingKeeper.AccountIsLiquidStakingProvider(ctx, icaAccountAddress), "ICA module account")
+}
+
+// Helper function to clear the Bonded pool balances before a unit test
+func clearPoolBalance(t *testing.T, app *simapp.SimApp, ctx sdk.Context) {
+	bondDenom := app.StakingKeeper.BondDenom(ctx)
+	initialBondedBalance := app.BankKeeper.GetBalance(ctx, app.AccountKeeper.GetModuleAddress(types.BondedPoolName), bondDenom)
+
+	err := app.BankKeeper.SendCoinsFromModuleToModule(ctx, types.BondedPoolName, minttypes.ModuleName, sdk.NewCoins(initialBondedBalance))
+	require.NoError(t, err, "no error expected when clearing bonded pool balance")
+}
+
+// Helper function to fund the Bonded pool balances before a unit test
+func fundPoolBalance(t *testing.T, app *simapp.SimApp, ctx sdk.Context, amount sdk.Int) {
+	bondDenom := app.StakingKeeper.BondDenom(ctx)
+	bondedPoolCoin := sdk.NewCoin(bondDenom, amount)
+
+	err := app.BankKeeper.MintCoins(ctx, minttypes.ModuleName, sdk.NewCoins(bondedPoolCoin))
+	require.NoError(t, err, "no error expected when minting")
+
+	err = app.BankKeeper.SendCoinsFromModuleToModule(ctx, minttypes.ModuleName, types.BondedPoolName, sdk.NewCoins(bondedPoolCoin))
+	require.NoError(t, err, "no error expected when sending tokens to bonded pool")
+}
+
+// Tests CheckExceedsGlobalLiquidStakingCap
+func TestCheckExceedsGlobalLiquidStakingCap(t *testing.T) {
```

Comment:
Could we make sure we have at least these three cases covered here:
1) No LSM tokenizations, only ICA account delegations. Exceeds cap.
2) ICA account delegations, only LSM tokenizations from non-module account. Exceeds cap.
3) Some LSM tokenizations, some ICA account delegations. Exceeds cap.
Should give us confidence both cases are individually covered and the sum too

Reply:
This tests [this](https://github.com/iqlusioninc/liquidity-staking-module/pull/61/files/744377e109d1430f13b3572bb6cf3a686f4d6df6#diff-fb1fb56df25f422dcd246c4517169542f2a04e040d8e39d261ee8ee1477bee46R55) function, and there's no differentiation between where the tokens came from

I do agree these are important scenarios to have covered, but I don't think this is the place to do it. It's important to keep the unit tests focused to the block of code that their testing.

Fwiw, there are more integration-like tests in `TestTokenizeSharesAndRedeemTokens` that cover these cases more closely